File: wwwroot/js/shader.js

function shaderAlert(msg) {
	console.log(msg);
	alert(msg);
}

function compileShader(gl, shaderURL, shaderString, shaderType) {
	var shader = gl.createShader(shaderType);
    gl.shaderSource(shader, shaderString);
    gl.compileShader(shader);

    if (!gl.getShaderParameter(shader, gl.COMPILE_STATUS) && !gl.isContextLost()) {
        shaderAlert("Error compiling shader: " + gl.getShaderInfoLog(shader));
		gl.deleteShader(shader);
        return null;
    }

	neg.shaderCache[shaderURL] = shader;
	console.log("shader: " + shaderURL + ": cached");
	
	return shader;
}

function linkProg(prog, gl, vertexShader, fragmentShader) {

	// link program
    var shaderProgram = gl.createProgram();
    gl.attachShader(shaderProgram, vertexShader);
    gl.attachShader(shaderProgram, fragmentShader);
    gl.linkProgram(shaderProgram);

    if (!gl.getProgramParameter(shaderProgram, gl.LINK_STATUS) && !gl.isContextLost()) {
        shaderAlert("Error linking program: " + gl.getProgramInfoLog(shaderProgram));
		return;
    }
	
	// save shader program
	prog.shaderProgram = shaderProgram;
}
		
function tryLinkProgram(prog) {
	
	if (!prog.vertexShader || !prog.fragmentShader) {
		// not ready
		return;
	}
	
	console.log("shader program: linking");
	linkProg(prog, gl, prog.vertexShader, prog.fragmentShader);

	shaderOngoingStop(prog);
}

function processVertexShader(opaque, response) {
	var prog = opaque;

	console.log(prog.vsFile + ": vertex shader: [" + response + "]");
	if (response == null) {
		shaderAlert("vertex shader: FATAL ERROR: could not load");
		shaderOngoingStop(prog);
		return;
	}
	prog.vertexShader = compileShader(gl, prog.vsFile, response, gl.VERTEX_SHADER);
	tryLinkProgram(prog);
}

function processFragmentShader(opaque, response) {
	var prog = opaque;
	
	console.log(prog.fsFile + ": fragment shader: [" + response + "]");
	if (response == null) {
		shaderAlert("fragment shader: FATAL ERROR: could not load");
		shaderOngoingStop(prog);
		return;
	}
	prog.fragmentShader = compileShader(gl, prog.fsFile, response, gl.FRAGMENT_SHADER);
	tryLinkProgram(prog);
}

function shaderOngoingStart(prog) {
}

function shaderOngoingStop(prog) {
	if ('shaderProgram' in prog) {
		console.log("shader.js: shader program: ready");
	}
	else {
		console.log("shader.js: shader program: failure");	
	}

	//neg.ongoingProgramLoads.splice(neg.ongoingProgramLoads.indexOf(prog), 1);
	
	// callback
	prog.callbackOnDone(prog);	
}

function fetchVertexShader(shaderURL, prog) {

	if (shaderURL in neg.shaderCache) {
		console.log("vertexShader: " + shaderURL + ": cache HIT");
		prog.vertexShader = neg.shaderCache[shaderURL];
		tryLinkProgram(prog);
		return;
	}

	console.log("vertexShader: " + shaderURL + ": cache MISS");
	fetchFile(shaderURL, processVertexShader, prog);
}

function fetchFragmentShader(shaderURL, prog) {

	if (shaderURL in neg.shaderCache) {
		console.log("fragmentShader: " + shaderURL + ": cache HIT");
		prog.fragmentShader = neg.shaderCache[shaderURL];
		tryLinkProgram(prog);
		return;
	}

	console.log("fragmentShader: " + shaderURL + ": cache MISS");
	fetchFile(shaderURL, processFragmentShader, prog);
}

function fetchProgramFromURL(vs, fs, callbackOnDone) {
	var prog = {};
	prog.vsFile = vs;
	prog.fsFile = fs;
	prog.callbackOnDone = callbackOnDone;
	
	//neg.ongoingProgramLoads = [];
	
	shaderOngoingStart(prog);
	
	fetchVertexShader(vs, prog);
	fetchFragmentShader(fs, prog);
}

File: wwwroot/dart/shader.dart
library shader;

import 'dart:html';
import 'dart:web_gl';
import 'dart:convert';
import 'dart:typed_data';
import 'dart:math' as math;

import 'package:vector_math/vector_math.dart';
import 'package:vector_math/vector_math_geometry.dart';
import 'package:vector_math/vector_math_lists.dart';
import 'package:game_loop/game_loop_html.dart';
import 'package:obj/obj.dart';

import 'camera.dart';
import 'texture.dart';
import 'asset.dart';
import 'logg.dart';
import 'selection.dart';

part 'buffer.dart';
part 'shader_tex.dart';
part 'picker.dart';
part 'solid.dart';

class ShaderProgram {
  Program program;
  RenderingContext gl;
  String programName;
  int a_Position;
  UniformLocation u_MV;
  UniformLocation u_P;
  bool shaderReady = false;

  List<Model> modelList = new List<Model>();

  ShaderProgram(this.gl, this.programName);

  /*
  void initContext(RenderingContext gl, Map<String,Texture> textureTable) {
  }
  */

  void getLocations() {
    a_Position = gl.getAttribLocation(program, "a_Position");
    u_MV = gl.getUniformLocation(program, "u_MV");
    u_P = gl.getUniformLocation(program, "u_P");
  }

  void fetch(Map<String, Shader> shaderCache, String vertexShaderURL,
      String fragmentShaderURL) {
    Shader compileShader(
        String shaderURL, String shaderSource, int shaderType) {
      Shader shader = gl.createShader(shaderType);
      gl.shaderSource(shader, shaderSource);
      gl.compileShader(shader);
      bool parameter =
          gl.getShaderParameter(shader, RenderingContext.COMPILE_STATUS);
      if (!parameter) {
        String infoLog = gl.getShaderInfoLog(shader);
        err("compileShader: compilation FAILURE: $shaderURL: info=$infoLog");
        if (gl.isContextLost()) {
          err("compileShader: compilation FAILURE: $shaderURL: info=$infoLog: context is lost");
        }
        return null;
      }

      if (shaderCache[shaderURL] != null) {
        err("compileShader: " +
            shaderURL +
            ": FIXME: overwriting shader cache");
      }
      shaderCache[shaderURL] = shader;

      return shader;
    }

    Shader vertexShader, fragmentShader;

    void tryLink() {
      if (vertexShader == null || fragmentShader == null) {
        // not ready to link
        return;
      }

      Program p = gl.createProgram();
      gl.attachShader(p, vertexShader);
      gl.attachShader(p, fragmentShader);
      gl.linkProgram(p);
      bool parameter = gl.getProgramParameter(p, RenderingContext.LINK_STATUS);
      if (!parameter) {
        String infoLog = gl.getProgramInfoLog(p);
        err("tryLink: shader program link FAILURE: $infoLog");
        if (gl.isContextLost()) {
          err("tryLink: shader program link FAILURE: $infoLog: context is lost");
        }
        return;
      }

      this.program = p;

      getLocations();

      shaderReady = true;
    }

    void fetchVertexShader() {
      String url = vertexShaderURL;

      var requestVert = new HttpRequest();
      requestVert.open("GET", url);
      requestVert.onLoad.listen((ProgressEvent e) {
        String response = requestVert.responseText;
        if (requestVert.status != 200) {
          err("vertexShader: url=$url: error: [$response]");
          return;
        }
        vertexShader =
            compileShader(url, response, RenderingContext.VERTEX_SHADER);
        tryLink();
      });
      requestVert.onError.listen((e) {
        err("vertexShader: url=$url: error: [$e]");
      });
      requestVert.send();
    }

    void fetchFragmentShader() {
      String url = fragmentShaderURL;

      var requestFrag = new HttpRequest();
      requestFrag.open("GET", url);
      requestFrag.onLoad.listen((ProgressEvent e) {
        String response = requestFrag.responseText;
        if (requestFrag.status != 200) {
          err("fragmentShader: url=$url: error: [$response]");
          return;
        }
        fragmentShader =
            compileShader(url, response, RenderingContext.FRAGMENT_SHADER);
        tryLink();
      });
      requestFrag.onError.listen((e) {
        err("fragmentShader: url=$url: error: [$e]");
      });
      requestFrag.send();
    }

    vertexShader = shaderCache[vertexShaderURL];
    if (vertexShader == null) {
      fetchVertexShader();
    }

    fragmentShader = shaderCache[fragmentShaderURL];
    if (fragmentShader == null) {
      fetchFragmentShader();
    }

    tryLink();
    // needed when both vertexShader and fragmentShader are found in cache
  }

  void addModel(Model newModel) {
    Model m = findModelByName(newModel.modelName);
    if (m != null) {
      err("Model.addModel: existing model modelName=${m.modelName}");
      return;
    }
    this.modelList.add(newModel);
  }

  Model findModelByName(String name) {
    Model mod;
    try {
      mod = modelList.firstWhere((m) {
        return m.modelName == name;
      });
    } on StateError {
      assert(mod == null); // not found
    }
    return mod;
  }

  Instance findInstance(String id) {
    Instance i;
    try {
      modelList.firstWhere((m) {
        i = m.findInstance(id);
        return i != null;
      });
    } on StateError {
      assert(i == null); // not found
    }
    return i;
  }

  void drawModels(GameLoopHtml gameLoop, Camera cam, Matrix4 pMatrix) {
    if (!shaderReady) {
      return;
    }

    gl.useProgram(program);
    gl.enableVertexAttribArray(a_Position);

    // send perspective projection matrix uniform
    gl.uniformMatrix4fv(this.u_P, false, pMatrix.storage);

    modelList.forEach((Model m) => m.drawInstances(gameLoop, this, cam));

    // clean up
    gl.bindBuffer(RenderingContext.ARRAY_BUFFER, null);
    gl.bindBuffer(RenderingContext.ELEMENT_ARRAY_BUFFER, null);

    //gl.disableVertexAttribArray(a_Position); // needed ??
  }

  void update(GameLoopHtml gameLoop) {
    modelList.forEach((m) => m.update(gameLoop));
  }
}
